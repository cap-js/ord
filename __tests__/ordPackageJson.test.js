const cds = require("@sap/cds");
const ord = require("../lib/ord");
const path = require("path");

jest.mock("../lib/date", () => ({
    getRFC3339Date: jest.fn(() => "2024-11-04T14:33:25+01:00")
}));

describe("Tests for default ORD document when .cdsrc.json is not present", () => {
    let csn;

    beforeAll(async () => {
        csn = await cds.load(path.join(__dirname, "bookshop", "srv"));
    });

    beforeEach(() => {
        cds.root = path.join(__dirname, "bookshop");
    });

    test("Successfully create ORD Documents with defaults", () => {
        const document = ord(csn);
        expect(document).toMatchSnapshot();
    });

    describe("apiResources", () => {
        // eslint-disable-next-line no-useless-escape
        const PACKAGE_ID_REGEX = /^([a-z0-9]+(?:[.][a-z0-9]+)*):(package):([a-zA-Z0-9._\-]+):(v0|v[1-9][0-9]*)$/;

        let document;

        beforeAll(() => {
            document = ord(csn);
        });

        test("partOfPackage values are valid ORD IDs ", () => {
            for (const apiResource of document.apiResources) {
                expect(apiResource.partOfPackage).toMatch(PACKAGE_ID_REGEX);
            }
        });

        test("The partOfPackage references an existing package", () => {
            for (const apiResource of document.apiResources) {
                expect(
                    document.packages.find(
                        (pck) => pck.ordId === apiResource.partOfPackage
                    )
                ).toBeDefined();
            }
        });
    });

    describe("eventResources", () => {
        // eslint-disable-next-line no-useless-escape
        const GROUP_ID_REGEX = /^([a-z0-9-]+(?:[.][a-z0-9-]+)*):([a-zA-Z0-9._\-/]+):([a-z0-9-]+(?:[.][a-z0-9-]+)*):(?<service>[a-zA-Z0-9._\-/]+)$/;

        let document;

        beforeAll(() => {
            document = ord(csn);
        });

        test("Assigned to exactly one CDS Service group", () => {
            for (const eventResource of document.eventResources) {
                expect(eventResource.partOfGroups.length).toEqual(1);
            }
        });

        test("The CDS Service Group ID includes the CDS Service identifier", () => {
            for (const eventResource of document.eventResources) {
                const [groupId] = eventResource.partOfGroups;
                expect(groupId).toMatch(GROUP_ID_REGEX);

                const match = GROUP_ID_REGEX.exec(groupId);
                if (match && match.groups?.service) {
                    let service = match.groups?.service;
                    if (service.startsWith("undefined"))
                        service = service.replace("undefined.", "");
                    const definition = csn.definitions[service];
                    expect(definition).toBeDefined();
                    expect(definition.kind).toEqual("service");
                }
            }
        });
    });
});
