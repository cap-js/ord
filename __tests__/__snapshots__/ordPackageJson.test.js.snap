// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for default ORD document when .cdsrc.json is not present Successfully create ORD Documents with defaults 1`] = `
{
  "$schema": "https://sap.github.io/open-resource-discovery/spec-v1/interfaces/Document.schema.json",
  "apiResources": [
    {
      "apiProtocol": "odata-v4",
      "description": "AdminService",
      "entityTypeMappings": {
        "entityTypeTargets": [
          {
            "ordId": "sap.odm:entityType:test-from-extension:v1",
          },
        ],
      },
      "entryPoints": [
        "/odata/v4/admin",
      ],
      "extensible": {
        "supported": "yes",
      },
      "ordId": "customer.capirebookshopordsample:apiResource:AdminService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:AdminService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.edmx",
        },
      ],
      "shortDescription": "short description for test AdminService",
      "title": "This is test AdminService title",
      "version": "2.0.0",
      "visibility": "private",
    },
    {
      "apiProtocol": "odata-v4",
      "description": "CatalogService",
      "entityTypeMappings": [
        {
          "entityTypeTargets": [
            {
              "ordId": "sap.odm:entityType:odm.bookshop.Authors:v1",
            },
          ],
        },
      ],
      "entryPoints": [
        "/browse",
      ],
      "extensible": {
        "supported": "no",
      },
      "ordId": "customer.capirebookshopordsample:apiResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "openapi-v3",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.oas3.json",
        },
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/xml",
          "type": "edmx",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.edmx",
        },
      ],
      "shortDescription": "CatalogService",
      "title": "CatalogService",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "description": "this is an application description",
  "eventResources": [
    {
      "description": "CAP Event resource describing events / messages.",
      "entityTypeMappings": {
        "entityTypeTargets": [
          {
            "ordId": "sap.odm:entityType:test-from-extension:v1",
          },
        ],
      },
      "extensible": {
        "supported": "yes",
      },
      "ordId": "customer.capirebookshopordsample:eventResource:AdminService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:AdminService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/AdminService.asyncapi2.json",
        },
      ],
      "shortDescription": "short description for test AdminService",
      "title": "This is test AdminService title",
      "version": "2.0.0",
      "visibility": "private",
    },
    {
      "description": "CAP Event resource describing events / messages.",
      "extensible": {
        "supported": "no",
      },
      "ordId": "customer.capirebookshopordsample:eventResource:CatalogService:v1",
      "partOfGroups": [
        "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      ],
      "partOfPackage": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "releaseStatus": "active",
      "resourceDefinitions": [
        {
          "accessStrategies": [
            {
              "type": "open",
            },
          ],
          "mediaType": "application/json",
          "type": "asyncapi-v2",
          "url": "/.well-known/open-resource-discovery/v1/api-metadata/CatalogService.asyncapi2.json",
        },
      ],
      "shortDescription": "CatalogService event resource",
      "title": "ODM capirebookshopordsample Events",
      "version": "1.0.0",
      "visibility": "public",
    },
  ],
  "groups": [
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:AdminService",
      "groupTypeId": "sap.cds:service",
      "title": "This is test AdminService title",
    },
    {
      "groupId": "sap.cds:service:customer.capirebookshopordsample:CatalogService",
      "groupTypeId": "sap.cds:service",
      "title": "Catalog Service",
    },
  ],
  "openResourceDiscovery": "1.9",
  "packages": [
    {
      "description": "Description for capire bookshop ord sample",
      "ordId": "customer.capirebookshopordsample:package:capirebookshopordsample:v1",
      "partOfProducts": [
        "customer:product:capire.bookshop.ord.sample:",
      ],
      "shortDescription": "Short description for capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:Customer:",
      "version": "1.0.0",
    },
  ],
  "policyLevel": "none",
  "products": [
    {
      "ordId": "customer:product:capire.bookshop.ord.sample:",
      "shortDescription": "Description for capire bookshop ord sample",
      "title": "capire bookshop ord sample",
      "vendor": "customer:vendor:customer:",
    },
  ],
}
`;
